src/Streamly/Internal/Data/Fold.hs

```diff
@@ -391,39 +390,31 @@ scan (Fold stepL initialL extractL) (Fold stepR initialR extractR) =
 
     where
 
+    runStep actionL sR = do
+        rL <- actionL
+        case rL of
+            Done bL -> do
+                rR <- stepR sR bL
+                case rR of
+                    Partial sR1 -> Done <$> extractR sR1
+                    Done bR -> return $ Done bR
+            Partial sL -> do
+                !b <- extractL sL
+                rR <- stepR sR b
+                return
+                    $ case rR of
+                        Partial sR1 -> Partial (sL, sR1)
+                        Done bR -> Done bR
+
     initial = do
-        r1 <- initialL
-        r2 <- initialR
-        case r1 of
-            Done b1 -> return $ case r2 of
-                Done b2 -> Done b2
-                Partial fs2 -> Partial (Nothing, Just fs2, b1)
-            Partial fs -> return $ case r2 of
-                Done b2 -> Done b2
-                Partial fs2 -> Partial (Just fs, Just fs2, undefined)
-
-    step (Just s1, s2, _) x = do
-        r1 <- stepL s1 x
-        case r1 of
-            Done b -> do
-                r2 <- stepR (fromJust s2) b
-                case r2 of
-                    Partial fs2 -> Done <$> extractR fs2
-                    Done b2 -> return $ Done b2
-            Partial fs -> do
-                !b <- extractL fs
-                r2 <- stepR (fromJust s2) b
-                return $ case r2 of
-                    Partial fs2 -> Partial (Just fs, Just fs2, undefined)
-                    Done b2 -> Done b2
-
-    step (Nothing, s2, b1) _ = do
-            r2 <- stepR (fromJust s2) b1
-            case r2 of
-                    Partial fs2 -> Done <$> extractR fs2
-                    Done b2 -> return $ Done b2
-
-    extract (_, s2, _) = extractR (fromJust s2)
+        r <- initialR
+        case r of
+            Partial sR -> runStep initialL sR
+            Done b -> return $ Done b
+
+    step (sL, sR) x = runStep (stepL sL x) sR
+
+    extract = extractR . snd
```

Comment:
extract for initial case (for stream scan very first value when both the Folds will be in the initial state).
extract should be something:
extract (Just s1, s2) = do        
        l1 <- extractL s1
        r2 <- stepR (fromJust s2) l1
        case r2 of
            Partial fs2 -> extractR fs2
            Done b2 -> return b2

Reply:
We have already done all this in `initial` when we called `runStep`. See this code in `runStep`:

```
Partial sL -> do
                !b <- extractL sL
                rR <- stepR sR b
                return
                    $ case rR of
                        Partial sR1 -> Partial (sL, sR1)
                        Done bR -> Done bR
```

All we need to do in extract is to extract `sR1`.

Reply:
kool.